compiler/il/OMRNode.cpp

```diff
@@ -2352,6 +2353,47 @@ OMR::Node::isNotCollected()
 
 
 bool
+OMR::Node::canBeInternalPtrAtObject()
+   {
+   TR::Compilation * comp = TR::comp();
+
+   if (!self()->hasPinningArrayPointer())
+      return false;
+   if (self()->isInternalPointer())
+      return true;
+
+   TR::NodeChecklist visited(comp);
+   TR::Node *curNode = self();
+   // Look for a node with symRef so that we can tell whether self() is originated from a collected reference
+   while (curNode)
+      {
+      // Come across a node that has already been checked
+      if (visited.contains(curNode))
+         return false;
+      if (curNode->isInternalPointer())
+         return true;
+      if (curNode->getOpCode().hasSymbolReference())
+         {
+         TR::Symbol *symbol = curNode->getSymbolReference()->getSymbol();
+         TR::ILOpCode op = curNode->getOpCode();
+         return op.isNew() ||
+                (op.getDataType() == TR::Address && op.isCall()) ||
+                symbol->isInternalPointer() ||
+                symbol->isCollectedReference();
+         }
+
+      visited.add(curNode);
+      if (curNode->getNumChildren() >= 1)
```

Comment:
Sorry, I just realized that we can't support aternary since the result is not known statically. Will add the assert.